### DISCUSSION TITLE
OData - to love it or to hate it ?  Struggling to use it.

### DISCUSSION THREAD

**Discussion by figuerres** [OData/AspNetCoreOData#1101] [General]
ok this is a bit of a vent sorry but i really have been struggling with some things and not been able to find the right answers..

1)  Swagger Open API -  every time i try to get OData and Swagger to work together its a pain.  when will we get some help with this ?
   some swagger ui packages do not see the odata at all.  others see it but do not show all the nice details like how to use filter and other options.  some tools can transform the metadata to swagger but they are not part of the default swagger tooling.

2)  RESTIER is an interesting package that can generate odata endpoints but its out of date and the way it creates the end points the standard swagger tools do not work with it. and with it beeping out of date i cant use it with current nuget packages.

3) routing and attributes and conventions oh my....   it seems like every version of dot net or odata packages changes the way we are supposed to write a controller and get it to work, it seems like a never ending trail of changes to the way to do it
by route attribute or by convention or by overload or by creating a function .....  what to return  SingleResult or ActionResult or ,,,,,,

i really wish that all the swagger and controller setup could just settle down and be a easy to setup as plain old rest api's are ...
or at least get us close to it.

today i was trying to setup a controller for a table that has two int keys , a  Get()  was fine but when i add Get(int x, inty) i get runtime errors about having the thing registered multiple times.....  
but another table with a single key  i can have
Get()
and 
Get(int x) 
and they work fine
but swagger shows the second get 2 times 

that is just one example of the stuff that just drives me mad ...... 

i love OData when it works!\

but in about 10 years of using it each time i got back to do a new project with the new versions it seems like the same problems keep coming back over and over....  the version chnages , the package name chnages but the pain is still there...

ok done with the venting.....

later i will try to just ask specific questions!

**Comment by wgibbons1** [comment]
Unfortunately I feel the same way. There seem to be quite a few bugs. On the plus side, the query operators are very good. I'm on the verge of deciding to have to only use it for GETs as all the issues I'm having are with POSTs.

**Comment by robertmclaws** [comment]
> ok this is a bit of a vent sorry but i really have been struggling with some things and not been able to find the right answers..

> 2. RESTIER is an interesting package that can generate odata endpoints but its out of date and the way it creates the end points the standard swagger tools do not work with it. and with it beeping out of date i cant use it with current nuget packages.

What about it is out of date? It's running on .NET 8.0, and it's using the latest ODL updates. Microsoft is still updating the 7.x branch of OData AspNetCore frequently, and it works just fine. Is there some feature you're missing from the OData 8.x branch that you need?

**Reply by figuerres** [reply]
@robertmclaws  first most of my complaints are directed at the  ot6her packages like the swagger packages and at the Microsoft OData libraries 
for Restier at the time i wrote this the 1.1 update has not been released so the picture went from say a C- grade to a solid B+ grade today.
i still want to see how V2 looks - then i might give it an A+ grade.

to the whole community: while the OData protocol has been fairly stable for years the dot net packages to hookup OData have had many changes that have made it difficult to set up.
 
right now all of the NuGet packages are pushing for V8 rather blindly and we have to reject some of the updates.
not all packages are able today to use the V8 numbered packages but so many packages are telling us to take the V8 versions...

**Comment by julealgon** [comment]
Just wanted to chime in to say that I share most of your concerns. I've been using OData ever since it was introduced into the .NET ecosystem with WCF Data Services and since then I believe I've used every flavor of it (although I've not yet tried RESTIER).

The experience feels subpar to me, compared to other official Microsoft libraries. It is clear that there are resourcing constraints around OData, although that does seem to have improved over the past couple years.

The libraries are always in "catch-up mode" either to newer patterns in the framework itself (like Minimal API), or to the OData SPEC itself, which to this day hasn't been fully implemented. Additionally, on this front, there are a LOT of bugs still sitting here in the repo as unsolved, things that, in some cases, have been broken for several years.

The documentation is not in a good place ATM. A lot of what exists is scattered in blog posts, in this repo, and most of it still points to old AspNet (non Core) OData in microsoft learn.

I've shared my frustration in the past with the team not pushing this library/framework enough... it's always a no-show at any major .NET events, framework launches, Microsoft Build, etc. The _vast_ majority of developers are completely unaware that OData is even a _thing_, all the know about is GraphQL. The "marketing" side of OData is in my opinion severely lacking.

I absolutely love OData and that's one of the reasons I want it to succeed, so seeing those (IMHO glaring) gaps from the outside is painful to me.
